### PR TITLE
Feat: Circuit for Removing the Quoted Printable Encoding in Email Body

### LINKED ISSUE #187: Make bodyparsing hashing work for more mail clients
Thanks to MV for the bug report.

Apparently apple mail automatically splits strings with the soft returns, and uses = for them, while also replacing the = in base64 padding with =3D

you can sidestep it by enforcing line lengths

Skiff and some other clients don't introduce any breaks, gmail web client introduces hard returns every 75 chars, apple mail introduces these weird soft returns every 75 chars. 

Seems like it would be really hard to account for all of these things in one circuit

Apple mail is specifically annoying because the simplest version of this email is just a base64 string of the encoded tx, and introducing random = padding chars really messes stuff up

### PR DESCRIPTION
**Pull request by shreyas-londhe**
## Description

This PR adds a helper circuit which proves the removal of soft line breaks introduced in Quoted Printable Encoding. Aims to resolve #187.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] I have discussed with the team prior to submitting this PR
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/circuits/helpers/remove-soft-line-breaks.circom [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    // Handle the last two characters
+    is_soft_break[encoded_length - 2] <== is_equals[encoded_length - 2] * is_cr[encoded_length - 2];
+    is_soft_break[encoded_length - 1] <== 0;
+
+    // Determine which characters should be zeroed
+    for (var i = 0; i < encoded_length; i++) {
+        if (i == 0) {
+            should_zero[i] <== is_soft_break[i];
+        } else if (i == 1) {
+            should_zero[i] <== is_soft_break[i] + is_soft_break[i-1];
```
[the remark this answers is not in the document]
Comment (shreyas-londhe) [reply packages/circuits/helpers/remove-soft-line-breaks.circom:88]: `is_soft_break` is set such that there is only 1 set bit in a window of 3, so I don't think there will be a case where the result will be non-binary. Can you think of a case?

File: packages/circuits/helpers/remove-soft-line-breaks.circom [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        r_dec[i] <== r_dec[i - 1] * r;
+    }
+
+    // Calculate rlc for processed
+    sum_enc[0] <== processed[0];
+    for (var i = 1; i < encoded_length; i++) {
+        sum_enc[i] <== sum_enc[i - 1] + r_enc[i] * processed[i];
+    }
+
+    // Calculate rlc for decoded
```
[the remark this answers is not in the document]
Comment (shreyas-londhe) [reply packages/circuits/helpers/remove-soft-line-breaks.circom:107]: I think the RLC takes care of that

File: packages/circuits/helpers/remove-soft-line-breaks.circom (hunk 1 above)
[the remark this answers is not in the document]
Comment (Divide-By-0) [reply packages/circuits/helpers/remove-soft-line-breaks.circom:72]: Oh I see, yeah I think you're right

**Comment by Divide-By-0** [comment]
Core circuit looks good to me. Can you make it an argument in the main EmailVerifier component arguments and if on, then apply it to the body string? Also, have you tested it on real emails at all? I'm not convinced they all use \r.

File: packages/circuits/helpers/remove-soft-line-breaks.circom [hunk 3]
```diff
@@ -0,0 +1,115 @@
+pragma circom 2.1.6;
+
+include "circomlib/circuits/comparators.circom";
+include "circomlib/circuits/mux1.circom";
+include "../utils/hash.circom";
+
+template RemoveSoftLineBreaks(maxLength) {
+    signal input encoded[maxLength];
+    signal input decoded[maxLength];
+    signal output is_valid;
```
Comment (saleel) [inline packages/circuits/helpers/remove-soft-line-breaks.circom:10]: We have been using camelCase for signal variables names. Would be nice to be consistent at-least for input/output signals.

Comment (shreyas-londhe) [reply packages/circuits/helpers/remove-soft-line-breaks.circom]: Thanks for pointing it out. Made the changes :)

File: packages/circuits/email-verifier.circom [hunk 4]
```diff
@@ -29,7 +30,7 @@ include "./utils/hash.circom";
 /// @input bodyHashIndex Index of the body hash `bh` in the emailHeader.
 /// @input precomputedSHA[32] Precomputed SHA-256 hash of the email body till the bodyHashIndex.
 /// @output pubkeyHash Poseidon hash of the pubkey - Poseidon(n/2)(n/2 chunks of pubkey with k*2 bits per chunk).
-template EmailVerifier(maxHeadersLength, maxBodyLength, n, k, ignoreBodyHashCheck) {
+template EmailVerifier(maxHeadersLength, maxBodyLength, n, k, ignoreBodyHashCheck, removeSoftLineBreaks) {
```
Comment (saleel) [inline packages/circuits/email-verifier.circom:33]: We can add comment for the new `removeSoftLineBreaks` param above like others.

**Review by saleel — APPROVED** [review]

**Pull request merged**